ci: move 4.19 complexity tests to tests-datapath-verifier GHA workflow

tklauser opened the pull request:
It seems the verifier issues with 4.19 are resolved when running complexity tests using LVH on that version. Let's cover it in the the GitHub actions workflow instead of Ginkgo tests.

Successful run with test commit: https://github.com/cilium/cilium/actions/runs/4489169184/jobs/7894672779

tklauser committed:
ci: move 4.19 complexity tests to tests-datapath-verifier GHA workflow

It seems the verifier issues with 4.19 are resolved when running
complexity tests using LVH on that version. Let's cover it in the GitHub
actions workflow instead of Ginkgo tests.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser commented on .github/workflows/tests-datapath-verifier.yaml:
```diff
@@ -150,8 +150,9 @@ jobs:
-        # TODO: add '4.19-main' once verifier issues are resolved on that version.
         include:
+          - kernel: '4.19-main'
+            ci-kernel: '419'
```
I'm not sure how version selection exactly works of the lvh action (/cc @brb), but it seems like we build images from latest 4.19.y stable branch: https://github.com/cilium/little-vm-helper-images/blob/8b568c7421308a200daa53cb595754a7f8e091a0/_data/kernels.json#L23-L26 Current 4.19-main is running 4.19.277:

```
root@complexity-test-4:~# uname -a
Linux complexity-test-4.19 4.19.277 #1 SMP Tue Mar 14 15:39:26 UTC 2023 x86_64 GNU/Linux
```

brb replied:
Every merged  little-vm-helper-images PR triggers a rebuild of the LVH VM images which might bump the kernel.

qmonnet reviewed (approved):
:rocket:

ti-mo commented:
/test

pchaigno reviewed (changes requested):
Sorry to be a bit of a pain, but this is highly suspicious to me. AFAIK, the complexity issue we were hitting was https://github.com/cilium/cilium/issues/17647#issuecomment-1014678098, which never got fixed on our side and was happening for more than a year.

Did it get fixed in the kernel without us knowing about it? Is there something else that fixed it very recently on our side? Or is this test actually broken again and not testing what we think?

I'll try to check if something recent on the kernel side could explain this.

ti-mo commented:
> Sorry to be a bit of a pain, but this is highly suspicious to me. AFAIK, the complexity issue we were hitting was [#17647 (comment)](https://github.com/cilium/cilium/issues/17647#issuecomment-1014678098), which never got fixed on our side and was happening for more than a year.
> 
> Did it get fixed in the kernel without us knowing about it? Is there something else that fixed it very recently on our side? Or is this test actually broken again and not testing what we think?
> 
> I'll try to check if something recent on the kernel side could explain this.

No problem at all, we need to know this stuff for sure. Take a look at this: https://github.com/cilium/cilium/actions/runs/4509588303/jobs/7939538892?pr=24538. This has much more detailed output and verifier stats for each tail call. It's based on the changes in this PR (adding 4.19 coverage) and loading all ELFs using Go.

I'll add a uname at the start of the test run so we know exactly what we're dealing with.

bimmlerd reviewed (approved):
CODEOWNERS change looks ok

ti-mo commented:
@pchaigno Does this really need to remain blocked on your investigation? Sounds like something that can be done on the side. The Go test code looks fine to me. If there are changes to be made to the feature flags, that should be done in a separate PR anyways.

ti-mo commented:
All tests green and https://github.com/cilium/cilium/actions/runs/4532665447/jobs/7984382192?pr=24538 (test run of https://github.com/cilium/cilium/pull/24538) shows that the largest program (`tail_handle_ipv4_cont`) of the largest bpf_lxc configuration is currently just within the 4.19 limit:

`verifier_test.go:194: tail_handle_ipv4_cont: processed 127506 insns (limit 131072), stack depth 264`

https://github.com/cilium/cilium/issues/17647 mentions a tail call split being merged close after the issue was reported. I see no reason to block this PR any further, so let's merge please.

pchaigno commented:
I ran a bunch of tests and it looks like it's working as expected. I don't have an explanation for why it was failing only a few weeks/months ago (no significant verifier or datapath changes were merged since then). It could just be small changes pushing us below & above the limit, though that's fairly rare with complexity issues as the complexity doesn't tend to increase/decrease linearly.

That all said, I also checked the workflow is passing and current master and it fails: https://github.com/cilium/cilium/actions/runs/4542358481/jobs/8005682276?pr=24599. That means if we were to merge, we would break CI :disappointed:

ldelossa reviewed (approved):
Approval contingent on: https://github.com/cilium/cilium/pull/24517#issuecomment-1486675725 passing.

ti-mo committed:
bpf: remove verifier-test.sh and check-complexity.sh

With all verifier-test.sh logic ported to a Go-driven test in package
test/verifier, this script is no longer being used.

The functionality of check-complexity.sh is provided out-of-the-box
by the Go verifier tests. Running `go test -v` displays verifier stats
for each tail call in Cilium.

Signed-off-by: Timo Beckers <timo@isovalent.com>

qmonnet reviewed (approved):
Putting my docs-structure hat on, and approving again

lmb committed:
test/verifier: don't dump full log into test output

The verifier is prone to produce several megabytes of log output
when things break. Writing this into the test output makes it difficult
to discern which tests actually fail.

Only dump the last 10 lines of verifier output to stderr and write
the full log to a file instead.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
bpf: disable cluster aware addressing in 4.19 complexity tests

The bpf_lxc complexity test fails on current 4.19 kernels due to a
backported kernel fix [0] that increases complexity across the board.
Our CI has so far not picked this up since the Jenkins VMs run a kernel
without the backport.

Reduce the tested feature set on affected 4.19 configurations so that
we can use the new LVH / GHA based CI infrastructure.

0: https://github.com/cilium/cilium/issues/17647#issuecomment-1014678098

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb commented:
/test-ci-verifier

lmb commented:
This is :heavy_check_mark:  on 4.19 now, with only cluster aware addressing disabled: https://github.com/cilium/cilium/actions/runs/4609988154/jobs/8147889395

zacharysarah reviewed (approved):
LGTM from a docs perspective.n 👍🏻
